url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/commands/auxiliarytools/TrajectoryRecorder.java

```diff
@@ -0,0 +1,677 @@
+		saveImageDialog.setPreferredSize(new Dimension(300, 75));
+		saveImageDialog.setContentPane(saveImageContainer);
+		
+		
+		//display the window
+		pack();
+		setVisible(true);
+		
+		
+	}
+	
+	
+	
+	private void handleKeyPressed(KeyEvent e){
+		
+		if(!selectedOnNewTrajectory){
+			return; //do not allow the user to change the state unless they are modifying the new trajectory space
+		}
+		
+		
+		String key = String.valueOf(e.getKeyChar());
+		
+
+		//otherwise this could be an action, see if there is an action mapping
+		String mappedAction = keyActionMap.get(key);
+		if(mappedAction != null){
+			
+			State curState = trajectory.getState(trajectory.numStates()-1);
+			
+			
+			//then we have a action for this key
+			//split the string up into components
+			String [] comps = mappedAction.split(" ");
+			String actionName = comps[0];
+			
+			//construct parameter list as all that remains
+			String params[];
+			if(comps.length > 1){
+				params = new String[comps.length-1];
+				for(int i = 1; i < comps.length; i++){
+					params[i-1] = comps[i];
+				}
+			}
+			else{
+				params = new String[0];
+			}
+			
+			Action action = domain.getAction(actionName);
+			if(action == null){
+				System.out.println("Unknown action: " + actionName);
+			}
+			else{
+				State nextState = action.performAction(curState, params);
+				GroundedAction ga = new GroundedAction(action, params);
+				trajectory.addActionStateTransition(ga, nextState);
+				
+				//now update the list
+				this.setIterationListData();
+				
+				//update the state
+				painter.updateState(nextState);
+				this.updatePropTextArea(nextState);
+				
+			}
+			
+		}
+		
+		
+		
+		
+		
+	}
+	
+	
+	private void resetNewTrajectory(){
+		if(!selectedOnNewTrajectory){
+			return;
+		}
+		
+		trainEl = null;
+		State initState = sg.generateState();
+		trajectory = new Trajectory(initState);
+		
+		painter.updateState(initState);
+		this.setIterationListData();
+		
+	}
+	
+	
+	
+	private void handleSaveButton(){
+		if(!selectedOnNewTrajectory){
+			return;
+		}
+		if(trajectory.numStates() == 1){
+			return; //don't save a trajectory with just an initial state
+		}
+		
+		//reset text
+		commandSaveTF.setText("Enter command of trajectory");
+		pathSaveTF.setText("enter save path");
+		
+		//pop open the dialog
+		saveDialog.pack();
+		saveDialog.setVisible(true);
+		
+	}
+	
+	private void handleSaveImage(){
+		saveImageDialog.pack();
+		saveImageDialog.setVisible(true);
+	}
+	
+	private void handleFinalSave(){
+		
+		//what is the save result?
+		String command = commandSaveTF.getText();
+		String fpath = pathSaveTF.getText();
+		
+		String path = directory + "/" + fpath;
+		
+		if(!path.endsWith(".txt")){
+			path = path + ".txt";
+		}
+		
+		TrainingElement te = new TrainingElement(command, trajectory);
+		String teSRep = parser.getStringOfTrainingElement(te);
+		
+		//write this string out
+		try{
+			BufferedWriter out = new BufferedWriter(new FileWriter(path));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 461, 462. The resource is closed at line 463. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).